include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,615 @@
+   * On receiving information on the target cell, the data is projected back to the
+   * quadrature points. In case local refinement is performed, we have to first
+   * project local DoF values of the parent element to each child.
+   *
+   * Since this class is templated by DATA type, the users DATA class has to
+   * be derived from the QPData class. In practice that amounts to implementing
+   * the following two functions shown below for a quadrature point data
+   * with 2 scalars :
+   * <code>
+   * struct MyQData : public QPData
+   * {
+   *   // a function to pack scalars into a vector
+   *   void pack_values(std::vector<double> &scalars) const
+   *   {
+   *     Assert (scalars.size()==2, ExcInternalError());
+   *     scalars[0] = value1;
+   *     scalars[1] = value2;
+   *   }
+   *
+   *   void unpack_values(const std::vector<double> &scalars)
+   *   {
+   *     Assert (scalars.size() ==2, ExcInternalError());
+   *     value1 = scalars[0];
+   *     value2 = scalars[1];
+   *   }
+   * };
+   * </code>
+   * This class can then be use as
+   * <code>
+   * distributed::QPManager<dim,MyQData> manager(fe,lhs,rhs);
+   * ...mark cells for refinement...
+   * manager.prepare_for_coarsening_and_refinement(tr,2); // there will be 2 values
+   * tr.execute_coarsening_and_refinement();
+   * ...initialize quadrature point data on new cells
+   * manager.interpolate(tr);
+   * </code>
+   * This approach can be extended to quadrature point data with Tensors of
+   * arbitrary order, although with a little bit more work in packing and
+   * unpacking of data.
+   *
+   * @note During mesh balancing children of a single cell can be attributed
+   * to different processors. As a corollary, their parent cells can exist on
+   * multiple processors.
+   *
+   * @note The functionality provided by this class can alternatively be achieved
+   * with the parallel::distributed::SolutionTransfer . However, that would require
+   * the following steps: (i) create an auxiliary DoFHandler with a (discontinuous Galerkin)
+   * FiniteElement which has enough components to represent all data stored at the
+   * quadrature points data. (ii) project the data to the FiniteElement space
+   * and thereby store results in vectors; (iii) use the parallel::distributed::SolutionTransfer
+   * to project those vectors to the new mesh; (iv) and finally project the
+   * data back to the quadrature points on the new mesh via FEValues class.
+   * The QPManager class aims at simplifying the whole process only by requiring
+   * that the quadrature point data class is derived from the QPData
+   * and thereby implements functions for packing and unpacking its data into a
+   * vector. Everything else will be done automatically.
+   *
+   * @author Denis Davydov, Jean-Paul Pelteret, 2016
+   */
+  template<int dim, typename DATA>
+  class QPManager : public QPManagerBase<typename parallel::distributed::Triangulation<dim>::cell_iterator, DATA>
```

Comment:
for hp::DoFHandler, the new manager class will be derived from `QPManagerBase` and will use some if its methods together with those `pack`/`unpack` from `QPData`.